compiler/src/dotty/tools/dotc/transform/MacrosSplitter.scala

```diff
@@ -0,0 +1,188 @@
+package dotty.tools.dotc
+package transform
+
+import core._
+import Decorators._
+import Types._
+import Contexts._
+import Symbols._
+import Constants._
+import Flags._
+import ast.Trees._
+import ast.{TreeTypeMap, tpd}
+import tasty.TreePickler.Hole
+import SymUtils._
+import NameKinds._
+
+import scala.collection.mutable
+import dotty.tools.dotc.core.StdNames._
+
+
+/** Translates macro implementation.
+ *  Checks that the phase consistency principle (PCP) holds on the body of a macro.
+ *
+ *
+ *  Transforms an inline macro definitions we assume that we have a single ~ directly as the RHS.
+ *  We will transform the definition from
+ *    ```
+ *    inline def foo[T1, ...](inline x1: X, ..., y1: Y, ....): Z = ~{ ... x1 ... '{ ... T1 ... x1 ... '(y) ... } }
+ *    ```
+ *  to
+ *    ```
+ *    inline def foo[T1, ...](inline x1: X, ..., y1: Y, ....): Z = ~foo$splice$1[T1, ...]('[T1], ..., x1, ..., '(y1), ...)
+ *
+ *    def foo$splice$1[T1, ...](T1$1: Type[T1], ..., x1$1: X, ..., y1: Expr[Y], ....): Z =
```

Comment:
What if the original splice is already of the form 

    ~bar(...)

In that case, we can simply forward to `bar`, no? Would it not be simpler to _require_ that all toplevel splices are of this form? 

If we do have a requirement like this, what is the added benefit of the phase?